### PR TITLE
FEM Beam reduced integration

### LINKED ISSUE #12320: [Problem] Pipe beam section cannot be used in FEM due to wrong element type
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

[Beam section](https://wiki.freecad.org/FEM_ElementGeometry1D) of the type Pipe is currently unusable because the following error appears:

```
 *ERROR reading *BEAM SECTION:
        *BEAM SECTION of type PIPE can
        only be used for B32R elements.
        Element            1  is not a B32R element.
```

A workaround is to manually change the element type when editing the input file (just add R to B32). 

On this occasion, the Reduced Integration option (B31 --> B31R and B32 --> B32R) could be added for beam elements since it's crucial when plasticity is involved: https://forum.freecad.org/viewtopic.php?t=61233

Also, once there's a fix for that, the obsolete `*BEAM GENERAL SECTION` keyword used by FreeCAD FEM should be replaced with `*BEAM SECTION`:

```
 *WARNING in calinput
          *BEAM GENERAL SECTION is obsolete
          please use *BEAM SECTION
```

### Full version info

```shell
OS: Windows 10 build 19045
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.35691 (Git)
Build type: Release
Branch: main
Hash: cc45b58dadae53d7a99be1ed7db25f6c93bb13d5
Python 3.10.13, Qt 5.15.8, Coin 4.0.2, Vtk 9.2.6, OCC 7.6.3
Locale: Polish/Poland (pl_PL)
```

### Subproject(s) affected?

FEM

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FEA-eng**
replaces #12444 and fixes #12320 (plus adds more versatility as mentioned in the comment under #12444)

This PR adds a new property BeamReducedIntegration to the CalculiX solver. If true (by default), beam elements B31 and B32 are replaced with B31R and B32R, respectively. This way it's possible to use the Pipe beam section and to obtain accurate results when plasticity is involved (mentioned in the comment under #12444).

### CODE REVIEW

**Comment by chennes** [comment]
@marioalexis84 can you please review?

File: src/Mod/Fem/femsolver/calculix/solver.py [hunk 1]
```diff
@@ -372,6 +372,16 @@ def add_attributes(obj, ccx_prefs):
...
 
+    if not hasattr(obj, "BeamReducedIntegration"):
+        obj.addProperty(
+            "App::PropertyBool",
+            "BeamReducedIntegration",
+            "Fem",
+            "Set to True to use beam elements with reduced integration"
+        )
+        red = ccx_prefs.GetBool("BeamReducedIntegration", True)
+        obj.BeamReducedIntegration = red
```
Comment (marioalexis84) [inline src/Mod/Fem/femsolver/calculix/solver.py:383]: You are trying to use the preference parameter `BeamReducedIntegration`, but it does not exist.
Usually you should associate a check box on the preferences page to that parameter.
If you don't want to relate the property to a preference parameter, just add the property without referencing a parameter and set the default value as `obj.BeamReducedIntegration = True`

**Review by marioalexis84 — CHANGES_REQUESTED** [review]

File: src/Mod/Fem/femsolver/calculix/write_mesh.py [hunk 2]
```diff
@@ -51,6 +51,10 @@ def write_mesh(ccxwriter):
         if ccxwriter.member.geos_fluidsection:
             meshtools.write_D_network_element_to_inputfile(ccxwriter.femmesh_file)
 
+        # Use reduced integration beam elements if this option is enabled in ccx solver settings
+        if ccxwriter.solver_obj.BeamReducedIntegration is True:
+            meshtools.beam_reduced_integration(ccxwriter.femmesh_file)
+
```
Comment (marioalexis84) [inline src/Mod/Fem/femsolver/calculix/write_mesh.py:57]: Remove `is True`.
Just `if ccxwriter.solver_obj.BeamReducedIntegration:`

File: src/Mod/Fem/femsolver/calculix/write_mesh.py [hunk 3]
```diff
@@ -69,6 +73,10 @@ def write_mesh(ccxwriter):
             # inpfile is closed
             meshtools.write_D_network_element_to_inputfile(ccxwriter.femmesh_file)
 
+        # Use reduced integration beam elements if this option is enabled in ccx solver settings
+        if ccxwriter.solver_obj.BeamReducedIntegration is True:
+            meshtools.beam_reduced_integration(ccxwriter.femmesh_file)
+
```
Comment (marioalexis84) [inline src/Mod/Fem/femsolver/calculix/write_mesh.py:79]: Remove `is True`.
Just `if ccxwriter.solver_obj.BeamReducedIntegration:`

**Comment by FEA-eng** [comment]
@marioalexis84 I made the requested changes. Thank you for the help with this PR.

File: src/Mod/Fem/femmesh/meshtools.py [hunk 4]
```diff
@@ -2508,4 +2508,21 @@ def compact_mesh(
     # may be return another value if the mesh was compacted, just check last map entries
     return (new_mesh, node_map, elem_map)
 
+# ************************************************************************************************
+def beam_reduced_integration(
```
Comment (chennes) [inline src/Mod/Fem/femmesh/meshtools.py:2512]: Could you please reformat to use four-space indentation? Thanks!

Comment (FEA-eng) [reply src/Mod/Fem/femmesh/meshtools.py:2512]: > Could you please reformat to use four-space indentation? Thanks!

Sure, done.

**Comment by marioalexis84** [comment]
@FEA-eng can you squash all commits into one?

**Comment by FEA-eng** [comment]
> @FEA-eng can you squash all commits into one?

Can it be done on the GitHub website (I don't use git locally) without having to create a new PR ? I can't find such an option anywhere.

**Comment by marioalexis84** [comment]
I have tested the code and it works well.
Some comments:
Why put the reduced integration option for beam section as a general property of the solver and instead not follow the logic used with fluid sections, for which their existence is checked in the analysis and only then the type of element is modified?
With this modification, an attempt is made to open, replace B32 with B32R and overwrite the .inp file in all cases, even if there are no beam sections in the analysis.

**Comment by marioalexis84** [comment]
> Can it be done on the GitHub website 

I don't know, I couldn't find it either.

**Comment by FEA-eng** [comment]
> I will squash on merge, so no need to do it manually unless you want it to be more than one commit. Please let me know, I will merge when I hear from you.

It can be a single commit if that reduces the mess caused by multiple changes, no problem.

**Comment by marioalexis84** [comment]
ok, so could it be a general property for reduced integration instead of `BeamReducedIntgration`? Or do you plan to include a reduced integration property for each element type that supports it? or property by item?
I ask this because once a property is added to a freecad object, future modifications must take into account past implementations in order to open old models without problems.
So it is best to decide now what you want to do.

**Comment by FEA-eng** [comment]
> ok, so could it be a general property for reduced integration instead of `BeamReducedIntgration`? Or do you plan to include a reduced integration property for each element type that supports it? or property by item? I ask this because once a property is added to a freecad object, future modifications must take into account past implementations in order to open old models without problems. So it is best to decide now what you want to do.

I think that it should have "Beam" in the name because otherwise, one could think that it applies to all element types which is not true. 

In the future, we can add e.g. `QuadReducedIntegration` (near future) and maybe also `HexReducedIntegration` (more distant future, less likely) in the same way if it's better for compatibility. Other types of elements don't use reduced integration so it would apply only to those. General `ReducedIntegration` would probably always be misleading but especially now.

**Comment by chennes** [comment]
@marioalexis84 is this good to go then? Discussion resolved?

**Review by marioalexis84 — APPROVED** [review]

**Comment by marioalexis84** [comment]
Yes, I forgot to press the approve button.

**Pull request merged**